src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public sealed class InstalledAndTransitivePackageFeed : PlainPackageFeedBase
+    {
+        private readonly IEnumerable<PackageCollectionItem> _installedPackages;
+        private readonly IEnumerable<PackageCollectionItem> _transitivePackages;
+        private readonly IPackageMetadataProvider _metadataProvider;
+
+        public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> installedPackages, IEnumerable<PackageCollectionItem> transitivePackages, IPackageMetadataProvider metadataProvider)
+        {
+            if (installedPackages == null)
+            {
+                throw new ArgumentNullException(nameof(installedPackages));
+            }
+
+            if (transitivePackages == null)
+            {
+                throw new ArgumentNullException(nameof(transitivePackages));
+            }
+
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+
+            _installedPackages = installedPackages;
+            _transitivePackages = transitivePackages;
+            _metadataProvider = metadataProvider;
+        }
+
+
+        /// <inheritdoc cref="IPackageFeed.ContinueSearchAsync(ContinuationToken, CancellationToken)" />
+        public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchAsync(ContinuationToken continuationToken, CancellationToken cancellationToken)
+        {
+            var searchToken = continuationToken as FeedSearchContinuationToken;
+            if (searchToken == null)
+            {
+                throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
+            }
+
+            // Remove foreign transitive packages
+            IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
```

Comment:
Foreign Transitive Package: A transitive package that is not brought by an installed package. Example

ProjectA has:
- PackageA that depends in T1, T2

ProjectB only has a project reference to ProjectA. Then:

ProjectA transitive packages with transitive origins:
- T1 (PackageA)
- T2 (PackageA) 

ProjectB transitive packages with transitive origins:
T1 (none)
T2 (none)
ProjectA (none)

In this case, ProjectB has foreign transitive dependencies.

Reply:
I agree with @jebriede. 

Please don't introduce a new term that doesn't have an obvious meaning. 

Just say "transitive packages from project references".

Reply:
I don't think that's right. 

The design doesn't seem to call that out, https://github.com/NuGet/Home/blob/dev/proposed/2021/TransitiveDependenciesInPMUI.md. 

cc @chgill-MSFT @jebriede 

Why are we exluding packages that are brought in through projects?

Reply:
That doesn't seem like a great idea. :( 

Transitive packages brought in through other projects *are* the current project's transitive dependencies. 
You can upgrade them, tweak etc. 
It also doesn't really match dotnet list package.

Reply:
Also sometimes, you may want a certain project to upgrade...ex. Executable vs Library.

Reply:
Team Review: Everyone should weigh in on whether or not transitive package references from project references should be included

Reply:
In my opinion, "transitive packages" in this context is every package that a particular project is consuming.  This should include packages flowing from projects.  Consider the scenario where my project has no package references but has one or more project references which pulls in packages.  This UI would show that my project has no top-level packages and no transitive packages right?  However, building my project would obviously show that I'm consuming packages.

Reply:
Thanks for all the perspectives so far everyone!

I chose to only show transitive packages coming from top-level packages rather than projects for simplicity - to make sure actions like "promoting/installing" a transitive package and scope are clear.

My thought process is that packages from referenced projects should be handled in the PMUIs for those respective projects. However, I am open to other perspectives and potentially changing that design decision if it makes sense.

**Question 1:** Let's say we do surface transitive packages from referenced projects, so Project B --> Project A ---> Package A.

I'm using in the PMUI for Project B and see that transitive Package A has a vulnerability.

I'm assume that directly installing a non-vulnerable version of Package A in Project A is the ideal fix rather than directly installing it in Project B. Is that the case or am I missing anything?

**Question 2:** What are examples of problems/confusions we might introduce by not surfacing transitive packages from referenced projects?

FYI @JonDouglas

Reply:
> I chose to only show transitive packages coming from top-level packages rather than projects for simplicity - to make sure actions like "promoting/installing" a transitive package and scope are clear.

Simplicity for promoting a package to top level a noble goal, but we this is a "show" first feature, not an action first one. 
You can always put a little indicator that tells people that they may want to upgrade the package in the originating project instead and not the  current one. 

I'd argue what's not simple is not showing all the transitive packages to begin with.

Reply:
@chgill-MSFT 

> I'm assume that directly installing a non-vulnerable version of Package A in Project A is the ideal fix rather than directly installing it in Project B. Is that the case or am I missing anything?

I think in that case, we should be smart enough to detect that the package actually came from Project A so we'd upgrade it in Project A.  At least if we show transitive packages, I'm more likely to see that I have a vulnerability somewhere in my dependency tree rather than having to go to each project to see what they have.

> Question 2: What are examples of problems/confusions we might introduce by not surfacing transitive packages from referenced projects?

One example is when Project B has no package references, only one or more project references.  I know that I'm using packages, but the UI shows me an empty list for "transitive packages".  But Solution Explorer shows me them, which I think is confusing.

Happy to chat about it so we're both on the same page!

Reply:
Assuming we can show all transitive dependencies, what would be the behavior on solution PM UI ?

Reply:
@chgill-MSFT let's chat tomorrow and potentially meet with everyone to review shortly.

Reply:
To echo @chgill-MSFT and also keep the project on-time with regards to delivery, here is another perspective.

1. The original design did take into account project references but was cut intentionally. The reason being is that the context of the user would be within the "project specific level PM UI" meaning the only context is the project packages and not it's project references. It would be confusing to have a transitive dependency be included from another project reference without a way to understand it's origin such as a UI affordance or hover experience. The goal with transitive packages is having a direct origin from their project's top-level package and not across project references to start. 
2. We should track a separate issue covering this design choice and with enough feedback post shipping this feature, we can iterate on including it with a new proposal specifically covering all flowing transitive packages across project boundaries. Including these packages in the list today would be confusing unless they had some special element inferring they came from another place outside the context of the project. @chgill-MSFT this would be a good issue to file.
3. Additionally, we may need to consider the current [.NET CLI experience](https://github.com/NuGet/Home/blob/dev/proposed/2020/Transitive-Dependencies.md#cli) to include origins which already support project references & the packages being brought in. Right now everything is just thrown into the list and it's not apparent where they came from. @chgill-MSFT this would also be a good issue to file for CLI <> IDE parity.

A lot of words to say these are all things we should consider while iterating further post release but shouldn't block on.

Reply:
Filled issue for point 2: https://github.com/NuGet/Home/issues/11624

Filled issue for point 3: https://github.com/NuGet/Home/issues/11625

Reply:
Can someone please follow up and update the spec please? 
What is the behavior when a transitive package is coming through both package and project roots?

Reply:
> Can someone please follow up and update the spec please? 

Followed up with spec update PR: https://github.com/NuGet/Home/pull/11640

> What is the behavior when a transitive package is coming through both package and project roots?

It will show the transitive package with project-resolved version. Example:

Suppose you have a solution with two projects:

- ProjectA:
  * ProjectB
  * Newtonsoft.Json.Bson 1.0.2 (has dependency on Newtonsoft.Json 12.0.1) 
- ProjectB:
  * Microsoft.TestPlatform.TestHost 17.2.0-preview-20220131-20 (has dependency on Newtonsoft.Json 9.0.1)

PM UI on ProjectA

![image](https://user-images.githubusercontent.com/1192347/156535354-b8ec86e9-5ba5-4879-ae87-4171ff743c08.png)

PM UI on ProjectB

![image](https://user-images.githubusercontent.com/1192347/156535436-b7d56194-7591-46f8-88c9-53c97df9c647.png)